File: swift/Sources/AetherProtocol/Security/PacketSigning.swift
// SPDX-License-Identifier: MIT

import Crypto
import Foundation

/// Packet signing service with nonce deduplication for replay prevention.
public actor PacketSigningService {
    /// Maximum age for cached nonces (5 minutes).
    private let maxNonceAgeSeconds: Int = 300

    /// Deduplication cache: (senderUhid, nonce) -> timestamp
    private var nonceCache: [String: (nonce: Data, timestamp: Date)] = [:]

    private let ed25519PrivateKey: Data
    private let ed25519PublicKey: Data

    public init(privateKey: Data, publicKey: Data) {
        self.ed25519PrivateKey = privateKey
        self.ed25519PublicKey = publicKey
    }

    /// Signs a mesh packet according to the Aether protocol.
    /// The signature covers: PacketNonce || TimestampMs || Type || SourceUhidLength || SourceUhid ||
    /// DestinationUhidLength || DestinationUhid || SHA256(Payload) || Ttl || Priority
    public func signPacket(_ packet: inout MeshPacket) throws {
        let signableData = try constructSignableData(packet)
        packet.signature = try Ed25519Service.sign(ed25519PrivateKey, signableData)
    }

    /// Verifies a packet signature and checks for replay attacks.
    public func verifyPacket(
        _ packet: MeshPacket,
        againstPublicKey publicKey: Data
    ) throws -> Bool {
        // Check nonce deduplication
        try checkNonceDuplicate(packet.sourceUhid, packet.packetNonce)

        // Construct and verify signature
        let signableData = try constructSignableData(packet)
        return Ed25519Service.verify(publicKey, signableData, packet.signature)
    }

    /// Gets the public key for this signing service.
    public func getPublicKey() -> Data {
        ed25519PublicKey
    }

    // MARK: - Private Methods

    private func constructSignableData(_ packet: MeshPacket) throws -> Data {
        var data = Data()

        // PacketNonce (8 bytes)
        data.append(packet.packetNonce)

        // TimestampMs (8 bytes, little-endian)
        var timestamp = packet.timestampMs.littleEndian
        data.append(withUnsafeBytes(of: &timestamp) { Data($0) })

        // Type (4 bytes, little-endian)
        var typeValue = Int32(packet.type.rawValue).littleEndian
        data.append(withUnsafeBytes(of: &typeValue) { Data($0) })

        // SourceUhidLength (4 bytes, little-endian)
        let sourceBytes = packet.sourceUhid.data(using: .utf8) ?? Data()
        var sourceLen = Int32(sourceBytes.count).littleEndian
        data.append(withUnsafeBytes(of: &sourceLen) { Data($0) })

        // SourceUhid (UTF-8)
        data.append(sourceBytes)

        // DestinationUhidLength (4 bytes, little-endian)
        let destBytes = packet.destinationUhid.data(using: .utf8) ?? Data()
        var destLen = Int32(destBytes.count).littleEndian
        data.append(withUnsafeBytes(of: &destLen) { Data($0) })

        // DestinationUhid (UTF-8)
        data.append(destBytes)

        // SHA256(Payload) (32 bytes)
        let payloadHash = Data(SHA256.hash(data: packet.payload))
        data.append(payloadHash)

        // Ttl (4 bytes, little-endian)
        var ttlValue = Int32(packet.ttl).littleEndian
        data.append(withUnsafeBytes(of: &ttlValue) { Data($0) })

        // Priority (4 bytes, little-endian)
        var priorityValue = Int32(packet.priority).littleEndian
        data.append(withUnsafeBytes(of: &priorityValue) { Data($0) })

        return data
    }

    private func checkNonceDuplicate(_ sourceUhid: String, _ nonce: Data) throws {
        let cacheKey = sourceUhid

        // Clean up old cache entries
        let now = Date()
        let threshold = now.addingTimeInterval(-TimeInterval(maxNonceAgeSeconds))
        for (key, value) in nonceCache {
            if value.timestamp < threshold {
                nonceCache.removeValue(forKey: key)
            }
        }

        // Check for duplicate nonce
        if let cachedEntry = nonceCache[cacheKey], cachedEntry.nonce == nonce {
            throw PacketSigningError.duplicateNonce(sourceUhid)
        }

        // Add to cache
        nonceCache[cacheKey] = (nonce: nonce, timestamp: now)
    }
}

public enum PacketSigningError: Error, Equatable {
    case duplicateNonce(String)
    case signatureFailed(String)
    case verificationFailed(String)
}

File: kotlin/src/main/kotlin/aethermesh/security/PacketSigning.kt
// SPDX-License-Identifier: MIT

package aethermesh.security

import aethermesh.protocol.MeshPacket
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.security.MessageDigest
import java.util.concurrent.ConcurrentHashMap

/**
 * Packet signing utilities with nonce deduplication.
 *
 * Constructs signable data matching the C# implementation exactly:
 * PacketNonce || TimestampMs || Type || SourceUhidLength || SourceUhid ||
 * DestinationUhidLength || DestinationUhid || SHA-256(Payload) || Ttl || Priority
 *
 * Nonce deduplication is keyed by `(sourceUhid, nonce)` so that:
 *  - a nonce collision across two different senders does NOT drop a
 *    legitimate packet, and
 *  - an attacker who pre-registers a nonce against a recipient cannot block
 *    the legitimate sender's first packet.
 *
 * Pre-2026-05-05 the cache used a `Pair<String, ByteArray>` key, but
 * `ByteArray.hashCode` / `equals` are identity-based — two distinct
 * `ByteArray` instances with identical bytes hashed differently and never
 * collided, so dedup silently never fired. Switched to a string key
 * `"<source>:<hex(nonce)>"` to match the C# reference (SourceUhid + ":" +
 * Convert.ToHexString(PacketNonce)).
 */
object PacketSigning {
    private val nonceDedupCache = ConcurrentHashMap<String, Long>()
    private const val MAX_PACKET_AGE_MS = 300_000L // 5 minutes
    private val HEX_CHARS = "0123456789ABCDEF".toCharArray()

    /**
     * Optional reputation hook.  When set, replay attempts and signature
     * failures are reported so the calling layer can down-score the offending
     * peer.  `null` (the default) disables all reputation side-effects without
     * changing validation semantics.
     */
    @Volatile
    var reputation: NodeReputationService? = null

    /** Convenience setter for injection from Java or builder-style callers. */
    fun setReputationService(service: NodeReputationService?) {
        reputation = service
    }

    /**
     * Constructs the signable data for a packet.
     *
     * Wire format (little-endian):
     *   PacketNonce (8 bytes)
     *   TimestampMs (8 bytes, int64)
     *   Type (4 bytes, int32)
     *   SourceUhidLength (4 bytes, int32)
     *   SourceUhid (UTF-8 bytes)
     *   DestinationUhidLength (4 bytes, int32)
     *   DestinationUhid (UTF-8 bytes)
     *   SHA-256(Payload) (32 bytes)
     *   Ttl (4 bytes, int32)
     *   Priority (4 bytes, int32)
     */
    fun constructSignableData(packet: MeshPacket): ByteArray {
        val sourceBytes = packet.sourceUhid.toByteArray(Charsets.UTF_8)
        val destBytes = packet.destinationUhid.toByteArray(Charsets.UTF_8)
        val payloadHash = computeSHA256(packet.payload)

        val buffer = ByteBuffer.allocate(
            8 +  // PacketNonce
                    8 +  // TimestampMs
                    4 +  // Type
                    4 + sourceBytes.size +  // SourceUhid
                    4 + destBytes.size +  // DestinationUhid
                    32 +  // SHA-256(Payload)
                    4 +  // Ttl
                    4    // Priority
        ).apply {
            order(ByteOrder.LITTLE_ENDIAN)
        }

        // PacketNonce
        buffer.put(packet.packetNonce)

        // TimestampMs
        buffer.putLong(packet.timestampMs)

        // Type
        buffer.putInt(packet.type.value.toInt())

        // SourceUhid length and data
        buffer.putInt(sourceBytes.size)
        buffer.put(sourceBytes)

        // DestinationUhid length and data
        buffer.putInt(destBytes.size)
        buffer.put(destBytes)

        // SHA-256(Payload)
        buffer.put(payloadHash)

        // Ttl
        buffer.putInt(packet.ttl)

        // Priority
        buffer.putInt(packet.priority.toInt())

        return buffer.array()
    }

    /**
     * Signs a packet using Ed25519.
     *
     * @param packet The packet to sign
     * @param privateKey 32-byte Ed25519 private key
     * @return 64-byte signature
     */
    fun signPacket(packet: MeshPacket, privateKey: ByteArray): ByteArray {
        val signableData = constructSignableData(packet)
        return Ed25519Service.sign(privateKey, signableData)
    }

    /**
     * Verifies a packet signature using Ed25519.
     *
     * @param packet The packet to verify
     * @param publicKey 32-byte Ed25519 public key
     * @return True if the signature is valid
     */
    fun verifyPacket(packet: MeshPacket, publicKey: ByteArray): Boolean {
        val signableData = constructSignableData(packet)
        val valid = Ed25519Service.verify(publicKey, signableData, packet.signature)
        if (!valid) reputation?.recordSignatureFailure(packet.sourceUhid)
        return valid
    }

    /**
     * Checks if a packet nonce has been seen before (replay prevention).
     * Returns true if the nonce is NEW (not a replay).
     *
     * Maintains a deduplication cache with a 5-minute TTL keyed by
     * `(sourceUhid, hex(nonce))` — see class-level docs for the rationale
     * vs. nonce-only keying.
     *
     * @param packet The packet to check
     * @return True if this is a new packet, false if it's a replay
     */
    fun isNewPacket(packet: MeshPacket): Boolean {
        val key = nonceKey(packet.sourceUhid, packet.packetNonce)
        val now = System.currentTimeMillis()

        // Clean old entries.
        nonceDedupCache.entries.removeAll { (_, timestamp) ->
            (now - timestamp) > MAX_PACKET_AGE_MS
        }

        // putIfAbsent is the atomic "first writer wins" check we want — if
        // it returns non-null, the nonce was already seen (replay).
        val isNew = nonceDedupCache.putIfAbsent(key, now) == null
        if (!isNew) reputation?.recordReplayAttempt(packet.sourceUhid)
        return isNew
    }

    /**
     * Composite dedup key: `"<sourceUhid>:<HEX(nonce)>"`. Hex is uppercase
     * to match `Convert.ToHexString` in the C# reference, so cross-language
     * peers logging the same key produce identical strings.
     */
    private fun nonceKey(sourceUhid: String, nonce: ByteArray): String {
        val sb = StringBuilder(sourceUhid.length + 1 + nonce.size * 2)
        sb.append(sourceUhid).append(':')
        for (b in nonce) {
            val v = b.toInt() and 0xFF
            sb.append(HEX_CHARS[v ushr 4])
            sb.append(HEX_CHARS[v and 0x0F])
        }
        return sb.toString()
    }

    /**
     * Test-only: clear the dedup cache. Production code never needs this —
     * entries TTL out automatically.
     */
    internal fun clearDedupCacheForTests() {
        nonceDedupCache.clear()
    }

    /**
     * Test-only: reset the reputation service reference to null so tests are
     * isolated from one another.
     */
    internal fun clearReputationServiceForTests() {
        reputation = null
    }

    /**
     * Computes SHA-256 hash of data.
     */
    private fun computeSHA256(data: ByteArray): ByteArray {
        val digest = MessageDigest.getInstance("SHA-256")
        return digest.digest(data)
    }
}
